### DISCUSSION TITLE
FreeRTOS CDC RP2040 examples skipped

### DISCUSSION THREAD

**Discussion by crstrand** [hathach/tinyusb#1348] [Q&A]
I have looked, but cannot find an explanation for why the RP2040 is skipped in the FreeRTOS CDC examples.
I would like to create a dual CDC device using FreeRTOS.

**Comment by hathach** [comment]
since rp2040 freeRTOS port is not available when it is added. If it is supported now, would you mind making an pr to update example ?

**Reply by crstrand** [reply]
Thanks for the reply.  I am trying to compile CDC using FreeRTOS on the RP2040.  It seems to build ok, but I can't seem to get it to work for me.  If there shouldn't be any reason it wouldn't work, has anyone been able to do it successfully?
I will hook up the Picoprobe and start in depth debugging if there is no reason it shouldn't work.

**Reply by hathach** [reply]
could you try to make and draft PR to update the cdc_msc_freertos example to support rp2040. When I have time or maybe other could help test/point out what is the missing piece.

**Reply by crstrand** [reply]
Could you point me to some documentation where I can become more familiar with the build environment for tinyusb so I can try to integrate that, RP2040SDK and FreeRTOS?

**Reply by hathach** [reply]
at this stage, tinyusb only has generic info doc, for implementation/porting detail, the code is the the docs

**Comment by kilograham** [comment]
make sure you use the correct RP2040 FreeRTOS port here https://github.com/FreeRTOS/FreeRTOS-Kernel/tree/main/portable/ThirdParty/GCC/RP2040

**Reply by hathach** [reply]
thank Graham for the suggestion, @crstrand please follow this port suggestion, the current default is probably the generic M0 one. Otherwise, when I could complete tinyusb porting for rp2040 pio-usb driver, I will try to get freertos example running.

**Reply by crstrand** [reply]
Using PicoSDK build environment, I have been able to get dual CDC working on the RP2040 using tinyUSB (as part of the Pico SDK) and FreeRTOS.  However, I don't understand the tinyUSB build environment enough to compile the RP2040 FreeRTOS examples that are skipped.
This is mostly due to my extremely limited understanding of cmake in general.

**Reply by hathach** [reply]
thanks @crstrand for the update, it is not a problem at all. I will update the example when I have time

**Comment by harbaum** [comment]
Since I am in the process of porting the support controller of [MiSTeryNano](https://github.com/harbaum/MiSTeryNano) from the BL616/M0S Dock to the Pico I needed Pico USB host and FreeRTOS. After finding these very useful hints here, it just worked out of the box with PIO-USB.
